### PR TITLE
EDSC-3612, make search results export asynchronous

### PR DESCRIPTION
**Pull request by DanielJDufour**
# Overview

### What is the feature?

Please summarize the feature or fix.

### What is the Solution?

Summarize what you changed.

### What areas of the application does this impact?

List impacted areas.

# Testing

### Reproduction steps

- **Environment for testing:**
- **Collection to test with:**

1. Step 1
2. Step 2...

### Attachments

Please include relevant screenshots or files that would be helpful in reviewing and verifying this change.

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: serverless-configs/aws-functions.yml [hunk 1]
```diff
@@ -130,6 +130,50 @@
     environment:
       updateOrderStatusStateMachineArn: ${self:resources.Outputs.UpdateOrderStatusWorkflow.Value}
 
+  exportSearchRequest:
+    handler: serverless/src/exportSearchRequest/handler.default
+    timeout: 300 # this lambda isn't restricted to 30s because it's triggered by SQS not API Gateway
```
Comment (DanielJDufour) [inline serverless-configs/aws-functions.yml:135]: move comment down

File: serverless-configs/aws-functions.yml [hunk 2]
```diff
@@ -130,6 +130,50 @@
...
+    memorySize: 256
```
Comment (DanielJDufour) [inline serverless-configs/aws-functions.yml:136]: keep memory default

File: serverless-configs/aws-resources.yml [hunk 3]
```diff
@@ -1,4 +1,27 @@
...
+          - "*"
+          AllowedHeaders:
+          - "*"
+
+  SearchExportQueue:
+    Type: 'AWS::SQS::Queue'
+    Properties:
+      QueueName: ${self:custom.siteName}-SearchExportQueue
+      ReceiveMessageWaitTimeSeconds: 20
+      VisibilityTimeout: 300
```
Comment (DanielJDufour) [inline serverless-configs/aws-resources.yml:23]: make larger than lambda timeout

File: serverless/src/edlOptionalAuthorizer/handler.js [hunk 4]
```diff
@@ -31,8 +31,7 @@ const edlOptionalAuthorizer = async (event) => {
   if (!jwtToken || jwtToken === '') {
     const authOptionalPaths = [
       '/autocomplete',
-      '/opensearch/granules',
-      '/collections/export'
```
Comment (DanielJDufour) [inline serverless/src/edlOptionalAuthorizer/handler.js:35]: put /collections/export back in

File: serverless/src/exportSearchRequest/__tests__/handler.test.js [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+const AWS_TEST_REGION = 'us-east-1'
+const SQS_TEST_PORT = 9324
+const SQS_TEST_HOST = `0.0.0.0:${SQS_TEST_PORT}`
+const SQS_TEST_QUEUE_NAME = 'SEARCH_EXPORT_REQUEST_TEST_QUEUE'
+const SQS_TEST_ENDPOINT = `http://${SQS_TEST_HOST}`
+const MOCK_USER_ID = 1234
+const MOCK_KEY = '00000000-0000-0000-0000-000000000000' // see /__mocks__/crypto.js
+
+// over-rides randomUUID to return the MOCK_KEY
+jest.mock('crypto')
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/__tests__/handler.test.js:25]: remove, for uuid package

File: serverless/src/exportSearchRequest/handler.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
+import { randomUUID } from 'crypto' // if we do 'node:crypto', jest.mock won't work
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/handler.js:1]: use older library

File: serverless/src/exportSearchRequest/handler.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    const { body, headers } = event
+
+    const { defaultResponseHeaders } = getApplicationConfig()
+
+    try {
+
+        const earthdataEnvironment = determineEarthdataEnvironment(headers)
+
+        const jwt = getJwtToken(event)
+        if (!jwt) throw Error("missing jwt")
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/handler.js:46]: make jwt optional

File: serverless/src/exportSearchRequest/handler.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+        const { id: userId } = getVerifiedJwtToken(jwt, earthdataEnvironment)
+
+        if (!userId) throw Error("failed getting userId from jwt")
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/handler.js:50]: auth is optional now

File: serverless/src/exportSearchRequest/handler.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+        const params = {
+            columns,
+            cursorpath,
+            format,
+            itempath,
+            query,
+            variables
+        }
+
+        const key = randomUUID()
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/handler.js:69]: use uuid lib

File: serverless/src/exportSearchRequest/handler.js [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            userId
+        }
+
+        const message = { params, extra }
+
+        const messageBody = JSON.stringify(message);
+
+        const dbConnection = await getDbConnection()
+
+        await dbConnection('exports').insert({
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/handler.js:88]: mock

File: serverless/src/exportSearchRequest/handler.js [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+            user_id: userId,
+            key,
+            state: "REQUESTED",
+            filename
+        })
+
+        const searchExportQueueUrl = getSearchExportQueueUrl()
+        console.log('searchExportQueueUrl:', searchExportQueueUrl)
+
+        await sqs.sendMessage({
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/handler.js:98]: mock it

File: serverless/src/exportSearchRequest/handler.js [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+            QueueUrl: searchExportQueueUrl,
+            MessageBody: messageBody
+        }).promise()
+        console.log('posted to search export queue')
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/handler.js:102]: prefix with lambda name, proper caps

Comment (DanielJDufour) [reply serverless/src/exportSearchRequest/handler.js:102]: and requestId!

File: serverless/src/exportSearchRequest/handler.js [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+
+        return {
+            isBase64Encoded: false,
+            statusCode: 200,
+            headers: {
+                ...defaultResponseHeaders,
+                'jwt-token': jwt
```
Comment (DanielJDufour) [inline serverless/src/exportSearchRequest/handler.js:109]: optiona

File: serverless/src/exportSearch/handler.js [hunk 14]
```diff
@@ -1,125 +1,189 @@
+import crypto from 'node:crypto'
```
Comment (DanielJDufour) [inline serverless/src/exportSearch/handler.js:1]: uuid lib instead

File: serverless/src/exportSearch/handler.js [hunk 15]
```diff
@@ -1,125 +1,189 @@
...
+ * Perform a loop through collection results and upload the full results in the requested format to S3
  * @param {Object} event Details about the HTTP request that it received
  */
 const exportSearch = async (event, context) => {
   // https://stackoverflow.com/questions/49347210/why-aws-lambda-keeps-timing-out-when-using-knex-js
   // eslint-disable-next-line no-param-reassign
   context.callbackWaitsForEmptyEventLoop = false
 
-  const { body, headers } = event
+  if (process.env.IS_OFFLINE || process.env.JEST_WORKER_ID) {
```
Comment (DanielJDufour) [inline serverless/src/exportSearch/handler.js:32]: re-evaluate

File: serverless/src/exportSearch/handler.js [hunk 16]
```diff
@@ -1,125 +1,189 @@
...
+  const { body } = sqsRecord
+
+  const { extra, params } = JSON.parse(body)
 
-  const { defaultResponseHeaders } = getApplicationConfig()
+  const { earthdataEnvironment, filename, key, jwt, requestId, userId } = extra
+  if (!filename) throw new Error("missing filename")
+  if (!key) throw new Error("missing key")
+  if (!requestId) throw new Error("missing requestId")
+  if (!userId) throw new Error("missing userId")
```
Comment (DanielJDufour) [inline serverless/src/exportSearch/handler.js:57]: optional

File: serverless/src/exportSearch/handler.js [hunk 17]
```diff
@@ -1,125 +1,189 @@
...
 
-  const { data, requestId } = JSON.parse(body)
+  const updateState = (state) => dbConnection('exports').where({ user_id: userId, key }).update({ state })
```
Comment (DanielJDufour) [inline serverless/src/exportSearch/handler.js:59]: userId optional

File: package.json [hunk 18]
```diff
@@ -31,7 +33,12 @@
     "cypress:open": "cypress open",
     "cypress:run": "cypress run -c video=false",
     "cypress:ci": "start-server-and-test start:ci http-get://localhost:8080 cypress:run",
-    "cypress:prepare-ci": "bin/cypress-prepare-ci"
+    "cypress:prepare-ci": "bin/cypress-prepare-ci",
+    "s3": "EXTRA_CORS_ALLOWED_ORIGINS=http://localhost:5000 MOTO_ALLOW_NONEXISTENT_REGION=True moto_server -H 0.0.0.0",
+    "s3:install": "brew install moto",
+    "s3:reset": "curl -X POST http://localhost:5000/moto-api/reset",
+    "sqs": "java -jar elasticmq-server-1.3.9.jar",
```
Comment (macrouch) [inline package.json:40]: I'd rather have sqs offline set up the same way as other projects our team manages, with the docker compose file.

I _hate_ the idea of running extra processes in order to run EDSC locally. [This PR for serverless offline sqs](https://github.com/CoorpAcademy/serverless-plugins/pull/117) shows that it is possible to disable sqs offline. So we'd get the benefit of queues locally, but only if the dev wants them. I don't think that is happening, but it is easy enough to open a new PR that is up to date and install the forked version of the plugin

**Comment by DanielJDufour** [comment]
superseded by https://github.com/nasa/earthdata-search/pull/1630

**Pull request closed without merging**